apps/android/app/src/main/java/chat/simplex/app/views/helpers/RecAndPlay.kt

```diff
@@ -0,0 +1,168 @@
+    rec.setOutputFile(filePath)
+    rec.prepare()
+    rec.start()
+    rec.setOnInfoListener { mr, what, extra ->
+      if (what == MEDIA_RECORDER_INFO_MAX_FILESIZE_REACHED) {
+        stop(recordingInProgress)
+        onStop()
+      }
+    }
+    return filePath
+  }
+
+  override fun stop(recordingInProgress: MutableState<Boolean>) {
+    if (!recordingInProgress.value) return
+    recordingInProgress.value = false
+    recorder?.metrics?.
+    runCatching {
+      recorder?.stop()
+    }
+    runCatching {
+      recorder?.reset()
+    }
+    runCatching {
+      // release all resources
+      recorder?.release()
+    }
+    recorder = null
+  }
+
+  override fun cancel(filePath: String, recordingInProgress: MutableState<Boolean>) {
+    stop(recordingInProgress)
+    runCatching { File(filePath).delete() }.getOrElse { Log.d(TAG, "Unable to delete a file: ${it.stackTraceToString()}") }
+  }
+}
+
+object AudioPlayer {
+  private val player = MediaPlayer().apply {
+    setAudioAttributes(
+      AudioAttributes.Builder()
+        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
+        .setUsage(AudioAttributes.USAGE_MEDIA)
+        .build()
+    )
+  }
+  private val helperPlayer: MediaPlayer =  MediaPlayer().apply {
+        setAudioAttributes(
+          AudioAttributes.Builder()
+            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
+            .setUsage(AudioAttributes.USAGE_MEDIA)
+            .build()
+        )
+  }
+  // Filepath: String, onStop: () -> Unit
+  val currentlyPlaying: MutableState<Pair<String, () -> Unit>?> = mutableStateOf(null)
+
+  fun start(filePath: String, seek: Int? = null, onStop: () -> Unit): Boolean {
+    if (!File(filePath).exists()) {
+      Log.e(TAG, "No such file: $filePath");
+      return false
+    }
+
+    val current = currentlyPlaying.value
+    if (current == null || current.first != filePath) {
+      player.reset()
+      // Notify prev audio listener about stop
+      current?.second?.invoke()
+      kotlin.runCatching {
+        player.setDataSource(filePath)
+      }.getOrElse { Log.e(TAG, it.stackTraceToString()); return false }
+      player.prepare()
+    }
+    if (seek != null) player.seekTo(seek)
+    player.start()
+    currentlyPlaying.value = filePath to onStop
+    return true
+  }
+
+  fun pause(): Int {
+    player.pause()
+    return player.currentPosition
+  }
+
+  fun stop() {
+    if (!player.isPlaying) return
+    // Notify prev audio listener about stop
+    currentlyPlaying.value?.second?.invoke()
+    currentlyPlaying.value = null
+    player.stop()
+  }
+
+  /**
+   * If player starts playing at 2637 ms in a track 2816 ms long, it will stop immediately after start but will not
+   *  change currentPosition, so it will not be equal to duration. However, it sets isPlaying to false.
+   *  Let's do it ourselves in order to prevent endless waiting loop
```

Comment:
is 2637 ms / 2816 ms an example or specific numbers?.. I'm trying to understand where it's implied in code/what fixes it

Reply:
This is just an example that I saw. Modified a comment to highlight it